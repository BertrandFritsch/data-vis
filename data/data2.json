[
  {
    "geo": "Austria",
    "population": 8772865
  },
  {
    "geo": "Belgium",
    "population": 11365834
  },
  {
    "geo": "Bulgaria",
    "population": 7101859
  },
  {
    "geo": "Croatia",
    "population": 4154213
  },
  {
    "geo": "Cyprus",
    "population": 854802
  },
  {
    "geo": "Czech Republic",
    "population": 10578820
  },
  {
    "geo": "Denmark",
    "population": 5748769
  },
  {
    "geo": "Estonia",
    "population": 1315635
  },
  {
    "geo": "Finland",
    "population": 5503297
  },
  {
    "geo": "France",
    "population": 67024459
  },
  {
    "geo": "Germany",
    "population": 82800000
  },
  {
    "geo": "Greece",
    "population": 10757293
  },
  {
    "geo": "Hungary",
    "population": 9797561
  },
  {
    "geo": "Ireland",
    "population": 4774833
  },
  {
    "geo": "Italy",
    "population": 60589445
  },
  {
    "geo": "Latvia",
    "population": 1950116
  },
  {
    "geo": "Lithuania",
    "population": 2847904
  },
  {
    "geo": "Luxembourg",
    "population": 590667
  },
  {
    "geo": "Malta",
    "population": 440433
  },
  {
    "geo": "Netherlands",
    "population": 17081507
  },
  {
    "geo": "Poland",
    "population": 37972964
  },
  {
    "geo": "Portugal",
    "population": 10309573
  },
  {
    "geo": "Romania",
    "population": 19638309
  },
  {
    "geo": "Slovakia",
    "population": 5435343
  },
  {
    "geo": "Slovenia",
    "population": 2065895
  },
  {
    "geo": "Spain",
    "population": 46528966
  },
  {
    "geo": "Sweden",
    "population": 9995153
  },
  {
    "geo": "United Kingdom",
    "population": 65808573
  }
]
