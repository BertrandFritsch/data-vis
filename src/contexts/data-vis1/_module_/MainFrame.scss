
.data-vis-container {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 1000px;
  margin: 0 auto;

  .data-vis-header {
    margin-top: 20px;
    display: flex;
    flex-direction: column;
    p {
      text-align: center;
    }

    .data-vis-header-1 {
      font-size: 2rem;
    }

    .data-vis-header-2 {
      font-size: 1.5rem;
    }
  }

  .data-vis-body {
    flex: 1;
    display: flex;
    min-height: 500px;

    .data-vis-chart {
      flex: 1;
      align-self: stretch;
      margin: 100px 0 100px 100px;
      min-width: 100px;

      svg {

        .data-vis-line {
          fill: none;
          stroke-linejoin: round;
          stroke-linecap: round;
          stroke-width: 1.5;

          &.data-vis-line-over {
            stroke-width: 2.5;
          }
        }

        .data-vis-axis-y {
          .domain {
            display: none;
          }
          .tick line {
            stroke: #767676;
          }

          .tick:not(.data-vis-axis-y-zero):not(:first-of-type) line {
            stroke-dasharray: 2, 2;
          }

          .data-vis-axis-y-zero line {
            stroke-width: 2;
          }
        }

        text {
          font-size: .75rem;
          font-family: ibm-plex-sans;
        }
      }
    }

    .data-vis-tools {
      align-self: center;

      margin: 100px 50px;
      display: flex;
      flex-direction: column;

      .bx--toggle__label {
        width: 10rem;
        margin: .5rem;
        font-size: .875rem;

        .bx--toggle__appearance {
          font-size: .875rem;
          margin-right: .5rem;
        }
      }

      .data-vis-container-place-over {
        font-weight: 600;
      }
    }
  }

  .data-vis-footer {
    display: flex;
    flex-direction: column;
    margin: 10px;
    p {
      text-align: center;
      font-size: .75rem;
    }
  }
}
