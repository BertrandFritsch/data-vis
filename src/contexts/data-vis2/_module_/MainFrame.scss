
.data-vis-container {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 1000px;
  margin: 0 auto;

  .data-vis-header {
    margin-top: 20px;
    display: flex;
    flex-direction: column;
    p {
      text-align: center;
    }

    .data-vis-header-1 {
      font-size: 2rem;
    }

    .data-vis-header-2 {
      font-size: 1.5rem;
    }
  }

  .data-vis-body {
    flex: 1;
    display: flex;
    height: 500px;
    width: 500px;
    align-self: center;

    .data-vis-chart {
      flex: 1;
      align-self: stretch;
      margin: 20px;

      svg {

        circle {
          fill: #6898cf;

          &:hover {
            fill: #5b7faf;
          }
        }

        text {
          font-size: .2rem;
          font-family: ibm-plex-sans;
          text-anchor: middle;
          dominant-baseline: central;
          fill: white;
          pointer-events: none;
        }
      }
    }

    .data-vis-tooltip {
      opacity: 0;
      position: absolute;
      pointer-events: none;
      transition: opacity 300ms;

      &.data-vis-tooltip--hover {
        opacity: .9;
      }

      div {
        font-weight: 100;
        font-size: .75rem;
        background-color: black;
        color: white;
        text-align: left;
        padding: 10px;
        border: 0;
        border-radius: 2px;

        span {
          line-height: 1.25rem;
        }
      }
    }
  }

  .data-vis-footer {
    display: flex;
    flex-direction: column;
    margin: 10px;
    p {
      text-align: center;
      font-size: .75rem;
    }
  }
}
