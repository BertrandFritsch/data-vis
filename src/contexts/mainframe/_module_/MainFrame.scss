
.data-vis-container {
  display: flex;
  flex-direction: column;
  height: 100%;

  .data-vis-header {
    height: 50px;
  }

  .data-vis-body {
    flex: 1;
    padding: 100px;

    svg {
      height: 100%;
      width: 100%;
    }

  }
}