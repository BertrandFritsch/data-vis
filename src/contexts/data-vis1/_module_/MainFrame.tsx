import * as React from 'react';
import './MainFrame.scss';
import * as d3 from 'd3';
import { PlaceProperties, TemperatureByDate, TemperaturesByPlace, WorldPlace } from './reducers';
import { setInArray } from '../../../helpers';

export interface Props {
  data: TemperaturesByPlace[];
  minValue: number;
  maxValue: number;
  minDate: Date;
  maxDate: Date;
  placeProperties: PlaceProperties;
}

interface State {
  selectedPlaces: WorldPlace[];
}

interface ChartState {
  update: () => void;
  resize: () => void;
}

export default class MainFrame extends React.PureComponent<Props, State> {
  state: State;
  private chartState: ChartState = { update: () => undefined, resize: () => undefined };

  constructor(props: Props) {
    super(props);

    this.state = { selectedPlaces: props.data.find(d => d.place === 'Glob') ? [ 'Glob' ] : [] };
  }

  componentDidMount() {
    window.addEventListener('resize', this.windowResized);
  }

  componentDidUpdate() {
    this.chartState.update();
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.windowResized);
  }

  render() {
    return (
      <div className='data-vis-container'>
        <div className='data-vis-header'>
          <p className='data-vis-header-1'>Zonal Annual Mean Land-Ocean Temperature</p>
          <p className='data-vis-header-2'>{ `${ this.props.minDate.getFullYear() } - ${ this.props.maxDate.getFullYear() }` }</p>
        </div>
        <div className='data-vis-body'>
          <div className='data-vis-chart' tabIndex={ -1 } ref={ this.renderChart } />
          <div className='data-vis-tools'>
            {
              this.props.data.map(r =>
                <div key={ r.place } style={ { color: this.props.placeProperties [ r.place ].color } }>
                  <div className='bx--form-item'>
                    <input className='bx--toggle bx--toggle--small' id={ `toggle-${ r.place }` } type='checkbox'
                           aria-label='Label Name'
                           checked={ this.state.selectedPlaces.find(d => d === r.place) !== undefined }
                           onChange={ e => this.togglePlace(r.place, e.target.checked) }
                    />
                    <label className='bx--toggle__label' htmlFor={ `toggle-${ r.place }` }>
                      <span className='bx--toggle__appearance'>
                      </span>
                      { this.props.placeProperties [ r.place ].text }
                    </label>
                  </div>
                </div>
              )
            }
          </div>
        </div>
        <div className='data-vis-footer'>
          <p className='data-vis-footer-1'>GISTEMP Team, 2018: GISS Surface Temperature Analysis (GISTEMP). NASA Goddard
            Institute for Space Studies. Dataset accessed 2018-02-10 at https://data.giss.nasa.gov/gistemp/</p>
          <p className='data-vis-footer-2'>Hansen, J., R. Ruedy, M. Sato, and K. Lo, 2010: Global surface temperature
            change, Rev. Geophys., 48, RG4004, doi:10.1029/2010RG000345</p>
        </div>
      </div>
    );
  }

  private windowResized = () => this.chartState.resize();

  private renderChart = (elem: HTMLElement | null) => {
    if (elem !== null) {
      this.chartState = (() => {

        let clientRect = elem.getBoundingClientRect();
        const margin = { top: 20, right: 40, bottom: 20, left: 40 };

        const $svg = d3.select(elem)
                       .append<SVGSVGElement>('svg')
                       .attr('width', clientRect.width)
                       .attr('height', clientRect.height);

        const startDate = this.props.data[ 0 ].values[ 0 ].date;
        const endDate = this.props.data[ 0 ].values[ this.props.data[ 0 ].values.length - 1 ].date;

        const $gRoot = $svg.append<SVGGElement>('g')
                           .attr('transform', `translate(${ margin.left },${ margin.top })`);

        const x = d3.scaleTime()
                    .domain([ startDate, endDate ]);

        const y = d3.scaleLinear()
                    .domain([ this.props.minValue, this.props.maxValue ])
                    .nice();

        const line = d3.line<TemperatureByDate>()
                       .x(d => x(d.date))
                       .y(d => y(d.temperature));

        const nbXTicks = (nbIntervals: number) => {
          // replace the last automatic tick by the real last one
          const ticks = x.ticks(nbIntervals);
          ticks[ ticks.length - 1 ] = endDate;
          return ticks;
        };

        const xAxis = d3.axisBottom(x);

        $gRoot.append<SVGGElement>('g')
              .attr('class', 'data-vis-axis-x');

        const yFormat = d3.format('.1f');

        const yAxis = d3.axisLeft<number>(y)
                        .tickFormat(d => d === 0 ? '0' : `${ yFormat(d) }Â°`);

        $gRoot.append<SVGGElement>('g')
              .attr('class', 'data-vis-axis-y')
              .call(yAxis);

        $gRoot.append('text')
              .attr('class', 'data-vis-info')
              .attr('dy', -3)
              .text('Annual temperature change compared to the mean of the years from 1951 to 1980');

        return {
          update: () => {

            const width = clientRect.width - margin.left - margin.right;
            const height = clientRect.height - margin.top - margin.bottom;
            const nbIntervals = Math.max(Math.trunc(width / 50), 1);

            $svg.attr('width', clientRect.width)
                .attr('height', clientRect.height);

            x.rangeRound([ 0, width ]);
            y.rangeRound([ height, 0 ]);

            xAxis.tickValues(nbXTicks(nbIntervals));
            yAxis.tickSize(-(width + 20));

            $gRoot.select<SVGGElement>('.data-vis-axis-x')
                  .transition()
                  .attr('transform', `translate(0,${ height })`)
                  .call(xAxis);

            $gRoot.select<SVGGElement>('.data-vis-axis-y')
                  .transition()
                  .call(yAxis)
                  .call($g => {
                    $g.selectAll('.tick')
                      .attr('class', d => `tick ${ d === 0 ? 'data-vis-axis-y-zero' : '' }`)
                      .call($g1 => $g1.select('line')
                                      .attr('x1', -20))
                      .call($g1 => $g1.select('text')
                                      .attr('dx', -3)
                                      .attr('dy', -3));
                  });

            const $chartLines = $gRoot.selectAll<SVGGElement, TemperaturesByPlace>('.data-vis-group-line')
                                      .data(this.props.data.filter(d => this.state.selectedPlaces.find(p => p === d.place) !== undefined), d => d.place);

            $chartLines.exit()
                       .remove();

            $chartLines.enter()
                       .append<SVGGElement>('g')
                       .attr('class', 'data-vis-group-line')
                       .call($g => $g.append<SVGPathElement>('path')
                                     .attr('class', 'data-vis-line')
                                     .style('stroke', d => this.props.placeProperties [ d.place ].color)
                       )
                       .merge($chartLines)
                       .transition()
                       .call($g => $g.select('.data-vis-line')
                                     .attr('d', d => line(d.values))
                       );
          },

          resize: () => {
            clientRect = elem.getBoundingClientRect();
            this.chartState.update();
          }
        };
      })();
      this.chartState.update();
    }
  };

  private togglePlace(place: WorldPlace, checked: boolean) {
    this.setState({
      selectedPlaces: setInArray(this.state.selectedPlaces, p => p === place, checked ? place : undefined)
    });
  }
}
