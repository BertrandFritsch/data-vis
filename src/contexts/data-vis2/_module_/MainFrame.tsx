import * as React from 'react';
import './MainFrame.scss';
import * as d3 from 'd3';
import { Record } from './reducers';

export interface Props {
  data: Record[];
}

interface Node {
  data: Record;
  x: number;
  y: number;
}

interface State {
  nodes: Node[];
}

interface ChartState {
  update: () => void;
}

export default class MainFrame extends React.PureComponent<Props, State> {
  state: State;

  private chartState: ChartState = { update: () => undefined };

  constructor(props: Props) {
    super(props);

    this.state = {
      nodes: props.data.map(d => ({
        data: d,
        x: 50,
        y: 50
      }))
    };
  }

  render() {
    return (
      <div className='data-vis-container'>
        <div className='data-vis-header'>
          <p className='data-vis-header-1'>European Union Population</p>
          <p className='data-vis-header-2'>2017</p>
        </div>
        <div className='data-vis-body'>
          <div className='data-vis-chart' tabIndex={ -1 } ref={ this.renderChart } />
        </div>
        <div className='data-vis-footer'>
          <p className='data-vis-footer-1'>Source: https://ec.europa.eu/eurostat/</p>
          <p className='data-vis-footer-2'>Copyright: https://ec.europa.eu/info/legal-notice_en</p>
        </div>
      </div>
    );
  }

  private renderChart = (elem: HTMLElement | null) => {
    if (elem !== null) {
      this.chartState = (() => {

        const clientRect = elem.getBoundingClientRect();

        const scale = d3.scaleLinear()
                        .domain([ 0, 1 ])
                        .range([ 1, 20 ]);

        const $svg = d3.select(elem)
                       .append<SVGSVGElement>('svg')
                       .attr('width', clientRect.width)
                       .attr('height', clientRect.height)
                       .attr('viewBox', '0, 0, 100, 100');

        const $tooltipDiv = d3.select(elem)
                              .append('div')
                              .attr('class', 'data-vis-tooltip');

        $tooltipDiv.append('div');

        const $overallGroup = $svg.append<SVGGElement>('g');
        const $circlesData = $overallGroup.selectAll<SVGGElement, Node>('.data-vis-circle')
                                          .data(this.state.nodes, d => d.data.geo);

        const populationFormat = d3.format(',.3r');

        const $circleGroups = $circlesData.enter()
                                          .append<SVGGElement>('g')
                                          .attr('class', 'data-vis-group-circle')
                                          .each((d, i, g) => {
                                            const $g = d3.select(g [ i ]);
                                            const $circle = $g.append('circle')
                                                              .attr('r', scale(d.data.relativePopulation))
                                                              .on('mousemove', () => {
                                                                $tooltipDiv.classed('data-vis-tooltip--hover', true)
                                                                           .style('left', d3.event.pageX + 4 + 'px')
                                                                           .style('top', d3.event.pageY + 4 + 'px')
                                                                           .select('div')
                                                                           .html(`<span>${ d.data.geo }</span><br /><span>${ populationFormat(d.data.population)}</span>`);
                                                              })
                                                              .on('mouseout', () => {
                                                                $tooltipDiv.classed('data-vis-tooltip--hover', false);
                                                              });

                                            if (d.data.relativePopulation > .4) {
                                              $g.append('text')
                                                .text(d.data.geo);
                                            }
                                            else {
                                              $circle.append('title')
                                                     .text(d.data.geo);
                                            }
                                          })
                                          .merge($circlesData);

        d3.forceSimulation(this.state.nodes)
          .force('collide', d3.forceCollide<Node>().radius(d => scale(d.data.relativePopulation * 1.05)).strength(.75))
          .force('x', d3.forceX(50))
          .force('y', d3.forceY(50))
          .on('tick', () => {
            $circleGroups.each(
              (d, i, g) => {
                d.x = Math.min(Math.max(scale(d.data.relativePopulation), d.x), 100 - scale(d.data.relativePopulation));
                d.y = Math.min(Math.max(scale(d.data.relativePopulation), d.y), 100 - scale(d.data.relativePopulation));
                d3.select(g[ i ])
                  .attr('transform', `translate(${ d.x },${ d.y })`);
              }
            );
          });
        return {
          update: () => {
            // nothing to update
          }
        };
      })();
      this.chartState.update();
    }
  };
}
