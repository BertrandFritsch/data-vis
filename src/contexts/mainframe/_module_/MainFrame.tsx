import * as React from 'react';
import './MainFrame.scss';
import * as d3 from 'd3';
import { Record } from './reducers';

export interface Props {
  data: Record[];
}

export default class MainFrame extends React.PureComponent<Props> {
  private chartState: { update: () => void } = { update: () => undefined };

  render() {
    return (
      <div className='data-vis-container'>
        <div className='data-vis-header' />
        <div className='data-vis-body' tabIndex={ -1 } ref={ this.renderChart } />
      </div>
    );
  }

  private renderChart = (elem: HTMLElement | null) => {
    if (elem !== null) {
      this.chartState = (() => {

        const $svg = d3.select(elem)
                       .append<SVGSVGElement>('svg');

        const clientRect = $svg.node()!.getBoundingClientRect();

        const margin = { top: 20, right: 10, bottom: 20, left: 10 };
        const width = clientRect.width - margin.left - margin.right;
        const height = clientRect.height - margin.top - margin.bottom;

        const $gRoot = $svg.append<SVGGElement>('g')
                           .attr('transform', `translate(${ margin.left },${ margin.top })`);

        console.log(d3.extent<Record, Date>(this.props.data, d => new Date(d.Year, 0, 1)) as [ Date, Date ])

        const x = d3.scaleTime()
                    .rangeRound([ 0, width ])
                    .domain(d3.extent<Record, Date>(this.props.data, d => new Date(d.Year, 0, 1)) as [ Date, Date ]);

        const y = d3.scaleLinear()
                    .rangeRound([ height, 0 ])
                    .domain(d3.extent<Record, number>(this.props.data, d => d.Glob) as [ number, number ]);

        const line = d3.line<Record>()
                       .x(d => x(new Date(d.Year, 0, 1)))
                       .y(d => y(d.Glob));

        $gRoot.append<SVGGElement>('g')
              .attr('transform', `translate(0,${ height / 2 })`)
              .call(d3.axisBottom(x));
              // .select('.domain')
              // .remove();

        $gRoot.append<SVGGElement>('g')
              .call(d3.axisLeft(y))
              .append('text')
              .attr('fill', '#000')
              .attr('transform', 'rotate(-90)')
              .attr('y', 6)
              .attr('dy', '0.71em')
              .attr('text-anchor', 'end')
              .text('Price ($)');

        $gRoot.append('path')
              .datum(this.props.data)
              .attr('fill', 'none')
              .attr('stroke', 'steelblue')
              .attr('stroke-linejoin', 'round')
              .attr('stroke-linecap', 'round')
              .attr('stroke-width', 1.5)
              .attr('d', line);

        return {
          update: () => {

          }
        };
      })();
      this.chartState.update();
    }
  };
}
