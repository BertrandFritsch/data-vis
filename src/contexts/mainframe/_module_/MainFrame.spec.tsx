import { mount, shallow } from 'enzyme';
import * as React from 'react';
import * as d3 from 'd3';
import MainFrame, { Props } from './MainFrame';
import { data } from '../../../../data/data';

describe('<MainFrame />', () => {
  const getInitialProps = (): Props => ({
    data
  });

  let root: HTMLDivElement = document.createElement('div');

  beforeEach(() => {
    root = document.createElement('div');
    document.body.appendChild(root);
  });

  afterEach(() => {
    root.parentNode!.removeChild(root);
  });

  it('should render correctly', () => {

    const props = { ...getInitialProps() };

    const wrapper = shallow<Props>(<MainFrame { ...props } />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render the svg element', () => {

    const props = { ...getInitialProps() };

    const wrapper = mount<Props>(<MainFrame { ...props } />, { attachTo: root });

    expect(root).toMatchSnapshot();
  });
});
