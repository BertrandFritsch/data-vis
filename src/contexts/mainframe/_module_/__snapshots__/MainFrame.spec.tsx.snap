// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainFrame /> should render correctly 1`] = `
<div
  className="data-vis-container"
>
  <div
    className="data-vis-header"
  />
  <div
    className="data-vis-body"
    tabIndex={-1}
  />
</div>
`;

exports[`<MainFrame /> should render the svg element 1`] = `
<div>
  <div
    class="data-vis-container"
  >
    <div
      class="data-vis-header"
    />
    <div
      class="data-vis-body"
      tabindex="-1"
    >
      <svg>
        <g
          transform="translate(10,20)"
        >
          <g
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="middle"
            transform="translate(0,480)"
          >
            <path
              class="domain"
              d="M0.5,6V0.5H980.5V6"
              stroke="#000"
            />
            <g
              class="tick"
              opacity="1"
              transform="translate(0.5,0)"
            >
              <line
                stroke="#000"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="#000"
                y="9"
              >
                .880
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(73.5,0)"
            >
              <line
                stroke="#000"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="#000"
                y="9"
              >
                .890
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(146.5,0)"
            >
              <line
                stroke="#000"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="#000"
                y="9"
              >
                .900
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(219.5,0)"
            >
              <line
                stroke="#000"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="#000"
                y="9"
              >
                .910
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(293.5,0)"
            >
              <line
                stroke="#000"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="#000"
                y="9"
              >
                .920
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(366.5,0)"
            >
              <line
                stroke="#000"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="#000"
                y="9"
              >
                .930
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(439.5,0)"
            >
              <line
                stroke="#000"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="#000"
                y="9"
              >
                .940
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(512.5,0)"
            >
              <line
                stroke="#000"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="#000"
                y="9"
              >
                .950
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(585.5,0)"
            >
              <line
                stroke="#000"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="#000"
                y="9"
              >
                .960
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(658.5,0)"
            >
              <line
                stroke="#000"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="#000"
                y="9"
              >
                .970
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(731.5,0)"
            >
              <line
                stroke="#000"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="#000"
                y="9"
              >
                .980
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(804.5,0)"
            >
              <line
                stroke="#000"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="#000"
                y="9"
              >
                .990
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(878.5,0)"
            >
              <line
                stroke="#000"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="#000"
                y="9"
              >
                :02
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(951.5,0)"
            >
              <line
                stroke="#000"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="#000"
                y="9"
              >
                .010
              </text>
            </g>
          </g>
          <g
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="end"
          >
            <path
              class="domain"
              d="M-6,960.5H0.5V0.5H-6"
              stroke="#000"
            />
            <g
              class="tick"
              opacity="1"
              transform="translate(0,905.5)"
            >
              <line
                stroke="#000"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="#000"
                x="-9"
              >
                -40
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,826.5)"
            >
              <line
                stroke="#000"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="#000"
                x="-9"
              >
                -30
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,748.5)"
            >
              <line
                stroke="#000"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="#000"
                x="-9"
              >
                -20
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,669.5)"
            >
              <line
                stroke="#000"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="#000"
                x="-9"
              >
                -10
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,590.5)"
            >
              <line
                stroke="#000"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="#000"
                x="-9"
              >
                0
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,511.5)"
            >
              <line
                stroke="#000"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="#000"
                x="-9"
              >
                10
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,433.5)"
            >
              <line
                stroke="#000"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="#000"
                x="-9"
              >
                20
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,354.5)"
            >
              <line
                stroke="#000"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="#000"
                x="-9"
              >
                30
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,275.5)"
            >
              <line
                stroke="#000"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="#000"
                x="-9"
              >
                40
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,197.5)"
            >
              <line
                stroke="#000"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="#000"
                x="-9"
              >
                50
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,118.5)"
            >
              <line
                stroke="#000"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="#000"
                x="-9"
              >
                60
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,39.5)"
            >
              <line
                stroke="#000"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="#000"
                x="-9"
              >
                70
              </text>
            </g>
            <text
              dy="0.71em"
              fill="#000"
              text-anchor="end"
              transform="rotate(-90)"
              y="6"
            >
              Price ($)
            </text>
          </g>
          <path
            d="M0,740L7,669L15,661L22,740L29,803L37,834L44,826L51,850L59,748L66,677L73,873L80,771L88,795L95,818L102,826L110,755L117,700L124,677L132,810L139,716L146,653L154,692L161,795L168,866L176,936L183,803L190,755L197,897L205,921L212,960L219,921L227,936L234,866L241,850L249,708L256,669L263,850L271,897L278,779L285,755L293,787L300,740L307,795L314,763L322,795L329,740L336,653L344,748L351,748L358,866L366,692L373,661L380,716L388,810L395,692L402,740L410,700L417,598L424,606L431,614L439,527L446,488L453,511L461,480L468,386L475,488L483,614L490,614L497,661L505,661L512,732L519,645L527,582L534,527L541,685L549,700L556,748L563,567L570,543L578,567L585,614L592,551L600,567L607,543L614,748L622,669L629,630L636,606L644,645L651,543L658,567L666,661L673,582L680,472L687,653L695,598L702,677L709,449L717,535L724,456L731,370L739,330L746,488L753,346L761,464L768,496L775,441L783,323L790,275L797,362L804,244L812,260L819,409L826,401L834,338L841,228L848,315L856,212L863,94L870,260L878,260L885,157L892,94L900,102L907,157L914,47L921,87L929,71L936,165L943,79L951,31L958,118L965,94L973,71L980,0"
            fill="none"
            stroke="steelblue"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </g>
      </svg>
    </div>
  </div>
</div>
`;
